File: lintcode/32.minimum-window-substring.py
class Solution:
    """
    @param source : A string
    @param target: A string
    @return: A string denote the minimum window, return "" if there is no such a string
    """

    def minWindow(self, source, target):
        target_map = {}
        source_map = {}

        for c in target:
            if c not in target_map:
                target_map[c] = 0
            source_map[c] = 0
            target_map[c] += 1

        left = 0
        count = 0
        minD = len(source) + 1
        minLeft = -1
        for right in range(len(source)):
            ch = source[right]
            if ch not in target_map:
                continue

            source_map[ch] += 1

            if source_map[ch] <= target_map[ch]:
                count += 1

            while count == len(target):
                if right - left + 1 < minD:
                    minD = right - left + 1
                    minLeft = left

                removed_ch = source[left]
                left += 1
                if removed_ch not in target_map:
                    continue
                source_map[removed_ch] -= 1
                if source_map[removed_ch] < target_map[removed_ch]:
                    count -= 1

        if minLeft == -1:
            return ""
        else:
            return source[minLeft:minLeft + minD]


if __name__ == '__main__':
    s = Solution()

    print(s.minWindow('ADOBECODEBANC', 'ABC'))

File: lintcode/s/32.minimum-window-substring.cpp
#include <unordered_map>
#include <vector>
using namespace std;
class Solution {
public:
  /**
   * @param source : A string
   * @param target: A string
   * @return: A string denote the minimum window, return "" if there is no such
   * a string
   */
  string minWindow(string &source, string &target) {
    unordered_map<char, int> needed;
    unordered_map<char, int> contained;

    for (char ch : target) {
      needed[ch] += 1;
      contained[ch] = 0;
    }

    int numNeeded = target.size();

    int minStart = -1;
    int minLen = source.size() + 1;
    int count = 0;

    for (int start = 0, end = 0; end < source.size(); end += 1) {
      char currChar = source[end];
      if (needed.count(currChar)) {
        contained[currChar] += 1;

        if (contained[currChar] <= needed[currChar]) {
          count += 1;
        }

        while (count == numNeeded) {
          int len = end - start + 1;

          if (len < minLen) {
            minStart = start;
            minLen = len;
          }

          if (needed.count(source[start])) {
            contained[source[start]] -= 1;
            if (contained[source[start]] < needed[source[start]]) {
              count -= 1;
            }
          }
          start += 1;
        }
      }
    }

    return minStart == -1 ? "" : source.substr(minStart, minLen);
  }
};
